### DISCUSSION TITLE
Incremental Git Reconciliation

### DISCUSSION THREAD

**Discussion by dishang-trilogy** [fluxcd/flux2#2585] [Q&A] [answered]
Hi, 
We have a Github repo with thousands of CRDs, and changes get committed every few seconds, we have been using flux 1 which can do incremental sync on the new commits so the changes can be applied really fast within a few seconds. 
We are trying out upgrading to flux 2, but the reconciliation cycle seems to do full sync which takes a while. Is there some configuration that can allow fast incremental sync?

**Comment by dishang-trilogy** [comment]
Hi @stefanprodan @hiddeco , do you know if this is possible?

**Comment by kingdonb** [comment, marked as answer]
Flux v2 does a full sync; if you need many resources to be synced quickly, it's best to break them up into several different groups.

It sounds like you can accept the resources being synced piecemeal, then why not separate them into pieces? Flux v2 does not have an incremental apply, all the resources must be valid in order for the apply to succeed as it starts with a dry-run and commits the full change with a server-side apply that can also be configured to wait for `healthChecks`, but it can no longer be configured for partial applies when some resources are invalid.

It sounds like you may have a valid performance issue that has to be documented, and could be fixed, but we will need more details about the issue and how it behaves to say whether it can be fixed.

Server-side applies should be very fast, but they do block until they are finished, and can also be configured to wait. I don't know how it will perform applying thousands of CRDs at once, but I guess not good as it has to parse and apply all of them through a dry-run every time, Flux v2 is really designed not for giant monolithic workloads but for sharding for performance, through different Flux Kustomizations that apply the huge collection of resources in blocks of resources atomically in smaller groups.

The work is done incrementally by Source Controller, which caches the repo and pulls commits incrementally every interval or when triggered for sync. You will have one GitRepository but many Kustomization and this way it should be performant. It will not let errors loose with no accounting of them like was done in Flux v1, where when used without Kustomize any resource that fails to validate it simply doesn't get applied with the rest and error gets swallowed into the logs.

Please take a look at the [flux-e2e](https://fluxcd.io/docs/flux-e2e/#kustomize-controller-builds-and-validates-resources) guide for a more in-depth idea of what happens when Flux applies a directory of resources through the Kustomize controller now.
